media_utils/video_puppeteer.py

```diff
@@ -0,0 +1,142 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import re
+
+from marionette_driver import By, expected, Wait
+from marionette_driver.errors import TimeoutException, NoSuchElementException
+
+
+# Adapted from https://github.com/gavinsharp/aboutmedia/blob/master/chrome/content/aboutmedia.xhtml
+_debug_script = """var mainWindow = window.QueryInterface(Components.interfaces.nsIInterfaceRequestor)
+                       .getInterface(Components.interfaces.nsIWebNavigation)
+                       .QueryInterface(Components.interfaces.nsIDocShellTreeItem)
+                       .rootTreeItem
+                       .QueryInterface(Components.interfaces.nsIInterfaceRequestor)
+                       .getInterface(Components.interfaces.nsIDOMWindow);
+var tabbrowser = mainWindow.gBrowser;
+for (var i=0; i < tabbrowser.browsers.length; ++i) {
+  var b = tabbrowser.getBrowserAtIndex(i);
+  var media = b.contentDocumentAsCPOW.getElementsByTagName('video');
+  for (var j=0; j < media.length; ++j) {
+     var ms = media[j].mozMediaSourceObject;
+     if (ms) {
+       debugLines = ms.mozDebugReaderData.split(\"\\n\");
+       return debugLines;
+     }
+  }
+}"""
+
+
+def playback_started(video):
+    try:
+        current_time = video.player_current_time
+        return current_time > 0
+    except Exception as e:
+        print ('Got exception %s' % e)
+        return False
+
+
+class VideoException(Exception):
+    pass
+
+# Debug info looks something like:
+# Dumping data for reader d10c000:
+#  Dumping Audio Track Decoders: - mLastAudioTime: 9.386666
+#   Reader 0: d2df400 ranges=[(0.000000, 16.042666)] active=true size=196456
+#  Dumping Video Track Decoders - mLastVideoTime: 8.466791
+#	 Reader 2: 13f54400 ranges=[(8.008000, 16.016000)] active=true size=3741917
+#	 Reader 1: d2edc00 ranges=[(4.004000, 8.008000)] active=false size=1088692
+#	 Reader 0: d2eac00 ranges=[(2.002000, 4.004000)] active=false size=242367
+
+# One simple hack: There should be two active readers at any one time. If
+# there are not exactly two, something is horribly wrong.
+
+active_re = re.compile("active=true", re.DOTALL)
+
+
+def playback_done(video):
+    time_left = video.player_duration - video.player_current_time
+    if abs(time_left) < 2.0:
+        return True
+
+    # Now, parse out playing info. If there is no active audio stream or no
+    # active video stream, throw an exception.
+
+    debug_lines = video.player_get_debug_lines()
+    if debug_lines is None:
+        return True
+
+    count = 0
```

Comment:
`debug_lines` is a list of strings whose terminating newlines have been trimmed off. `findall` doesn't care about newlines, it will give you all non-overlapping matches in the string. Could there be more than one `active=true` per line?